src/rpc/blockchain.cpp

```diff
@@ -286,9 +287,13 @@ static RPCHelpMan waitfornewblock()
     NodeContext& node = EnsureAnyNodeContext(request.context);
     Mining& miner = EnsureMining(node);
 
-    auto block{CHECK_NONFATAL(miner.getTip()).value()};
+    BlockRef block{CHECK_NONFATAL(miner.getTip()).value()};
     if (IsRPCRunning()) {
-        block = timeout ? miner.waitTipChanged(block.hash, std::chrono::milliseconds(timeout)) : miner.waitTipChanged(block.hash);
+        std::optional<BlockRef> maybe_block{timeout ? miner.waitTipChanged(block.hash, std::chrono::milliseconds(timeout)) : miner.waitTipChanged(block.hash)};
+        // Return new block unless the node is shutting down
+        if (maybe_block) {
+            block = *CHECK_NONFATAL(maybe_block);
+        }
     }
```

Comment:
Given that "no tip" is expected here (unlikely but is not a "programming logic error"), maybe avoid `CHECK_NONFATAL()` and use `JSONRPCError`:

```cpp
std::optional<BlockRef> old_block{miner.getTip()};
std::optional<BlockRef> new_block;
if (IsRPCRunning()) {
    const uint256 h{old_block.has_value() ? old_block->hash : uint256::ZERO};
    new_block = timeout > 0 ? miner.waitTipChanged(h, std::chrono::milliseconds(timeout)) : miner.waitTipChanged(h);
}

if (!new_block.has_value()) {
    throw JSONRPCError(RPC_CLIENT_NOT_CONNECTED, "No tip within timeout or shutting down");
}
```

Reply:
I mean to avoid `CHECK_NONFATAL()` if the tip is null and use `JSONRPCError` instead.

If the tip is null at the start, then wait for it to be set, if not set within the timeout, then `JSONRPCError`.

Reply:
I ended up with something similar...